src/offset/local/unix.rs

```diff
@@ -12,33 +12,59 @@ use std::sync::Once;
 
 use super::tz_info::TimeZone;
 use super::{DateTime, FixedOffset, Local, NaiveDateTime};
-use crate::Utc;
+use crate::{Datelike, LocalResult, Utc};
 
 pub(super) fn now() -> DateTime<Local> {
-    let now = Utc::now();
-    DateTime::from_utc(now.naive_utc(), offset(now.timestamp()))
+    let now = Utc::now().naive_utc();
+    DateTime::from_utc(now, offset(now, false).unwrap())
 }
 
-pub(super) fn naive_to_local(d: &NaiveDateTime, local: bool) -> DateTime<Local> {
-    let offset = match local {
-        true => offset(d.timestamp()),
-        false => FixedOffset::east(0),
-    };
-
-    DateTime::from_utc(*d - offset, offset)
+pub(super) fn naive_to_local(d: &NaiveDateTime, local: bool) -> LocalResult<DateTime<Local>> {
+    if local {
+        match offset(*d, true) {
+            LocalResult::None => LocalResult::None,
+            LocalResult::Ambiguous(early, late) => LocalResult::Ambiguous(
+                DateTime::from_utc(*d - early, early),
+                DateTime::from_utc(*d - late, late),
+            ),
+            LocalResult::Single(offset) => {
+                LocalResult::Single(DateTime::from_utc(*d - offset, offset))
+            }
+        }
+    } else {
+        LocalResult::Single(DateTime::from_utc(*d, offset(*d, false).unwrap()))
+    }
 }
 
-fn offset(unix: i64) -> FixedOffset {
+fn offset(d: NaiveDateTime, local: bool) -> LocalResult<FixedOffset> {
     let info = unsafe {
         INIT.call_once(|| {
             INFO = Some(TimeZone::local().expect("unable to parse localtime info"));
         });
         INFO.as_ref().unwrap()
     };
```

Comment:
Potentially this caching behavior should be configurable. In a long running application, it might be unexpected that even when the system timezone changes, the application still uses the old timezone